src/php/Command/ReplCommand.php

```diff
@@ -21,102 +20,123 @@ final class ReplCommand
     private EvalCompilerInterface $compiler;
     private ExceptionPrinterInterface $exceptionPrinter;
     private ColorStyleInterface $style;
+    private PrinterInterface $printer;
+
+    /** @var string[] */
+    private array $inputBuffer = [];
 
     public function __construct(
         ReplCommandIoInterface $io,
         EvalCompilerInterface $compiler,
         ExceptionPrinterInterface $exceptionPrinter,
-        ColorStyleInterface $style
+        ColorStyleInterface $style,
+        PrinterInterface $printer
     ) {
         $this->io = $io;
         $this->compiler = $compiler;
         $this->exceptionPrinter = $exceptionPrinter;
         $this->style = $style;
+        $this->printer = $printer;
     }
 
     public function run(): void
     {
         $this->io->readHistory();
-        $this->io->output($this->style->yellow("Welcome to the Phel Repl\n"));
-        $this->io->output('Type "exit" or press Ctrl-D to exit.' . "\n");
+        $this->io->output($this->style->yellow('Welcome to the Phel Repl' . PHP_EOL));
+        $this->io->output('Type "exit" or press Ctrl-D to exit.' . PHP_EOL);
 
         $this->loopReadLineAndAnalyze();
     }
 
-    /**
-     * @throws ExitException
-     */
     private function loopReadLineAndAnalyze(): void
     {
         while (true) {
             try {
-                if ($this->io->isBracketedPasteSupported()) {
-                    $this->io->output(self::ENABLE_BRACKETED_PASTE);
-                }
-                $input = $this->io->readline(self::PROMPT);
-                if ($this->io->isBracketedPasteSupported()) {
-                    $this->io->output(self::DISABLE_BRACKETED_PASTE);
-                }
-
-                try {
-                    $this->checkInputAndAnalyze($input);
-                } catch (ExitException $e) {
-                    break;
-                }
+                $this->addLineFromPromptToBuffer();
+                $this->checkExitInputBuffer();
+                $this->analyzeInputBuffer();
+            } catch (ExitException $e) {
+                break;
             } catch (Throwable $e) {
-                $this->io->output($e->getTraceAsString() . PHP_EOL);
+                $this->inputBuffer = [];
+                $this->io->output($this->style->red($e->getMessage() . PHP_EOL));
             }
         }
 
-        $this->io->output($this->style->yellow("Bye!\n"));
+        $this->io->output($this->style->yellow('Bye!' . PHP_EOL));
+    }
+
+    private function addLineFromPromptToBuffer(): void
+    {
+        if ($this->io->isBracketedPasteSupported()) {
+            $this->io->output(self::ENABLE_BRACKETED_PASTE);
+        }
+
+        $prompt = empty($this->inputBuffer) ? self::INITIAL_PROMPT : self::OPEN_PROMPT;
+        $input = $this->io->readline($prompt);
+
+        if ($this->io->isBracketedPasteSupported()) {
+            $this->io->output(self::DISABLE_BRACKETED_PASTE);
+        }
+
+        if ($input === null) {
```

Comment:
If `$input === null` the user pressed `CTRL+D`.
* If I'm in multiline mode I would expect to exit the multiline mode and start with a new prompt `>>> `
* If I'm not in multiline mode the REPL should exit.

Maybe this is implemented like this. But I can't see it.

The `exit` command is only handle in non multiline mode. 
* `>>> exit` will stop the REPL
* `... exit` will not stop the REPL. `exit` is then interpreted as a Phel symbol.

Reply:
Exactly. This is the intended behavior.

Reply:
I tested it now. But it is not working. My input was this:
```
>>> (+
... 1
...
```
The I pressed `CTRL+D` and this happens:
```
>>> (+
... 1
... ...
```

After pressing `CTRL+D` multiple times:
```
>>> (+
... 1
... ... ... ... ... ... ... ... ... ... ... ... ...
```

If i insert this (last insert was `2`, no `enter`)
```
>>> (+
... 1
... 2
```
and press `CTRL+D` nothing happens

Reply:
Right! I was using Ctrl+C all the time, so I didn't properly test this one 🙈 
It should be fixed now. Thanks for checking it.

Reply:
Actually, regarding your last scenario:
```
If i insert this (last insert was 2, no enter)

>>> (+
... 1
... 2
and press CTRL+D nothing happens
```
Yes, the control+d scenario is really tricky, because of how it's implemented by design.
`readline()` returns false (and we cast it to null, see ReplCommandSystemIo:28) when "there is no more data to read, then FALSE is returned." (https://php.net/manual/en/function.readline.php)

I think it's ok like it is right now. Ctrl+C will exit the execution for sure always :)
Anyway, if you have another idea or suggestion on how to do that, please, don't hesitate. Anyway, we can always do a follow up PR with some other improvements later on.

Reply:
Ah that is ok for me.